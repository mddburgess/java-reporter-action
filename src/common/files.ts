import * as glob from '@actions/glob';

export async function findFiles(searchPaths: string[]) {
    const globber = await glob.create(searchPaths.join('\n'));
    return await globber.glob();
}

export async function findFile(searchPath: string) {
    const paths = await findFiles([searchPath]);
    return paths[0];
}

export async function findRelativePath(searchPath: string) {
    const globber = await glob.create(searchPath);
    const searchRoot = globber.getSearchPaths()[0];
    const absolutePaths = await globber.glob();
    return absolutePaths[0] && absolutePaths[0].slice(searchRoot.length + 1);
}
