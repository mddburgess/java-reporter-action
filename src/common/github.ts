export interface Annotation {
    path: string,
    start_line: number,
    end_line: number,
    start_column?: number,
    end_column?: number,
    annotation_level: AnnotationLevel,
    title?: string,
    message: string,
    raw_details?: string
}

export type AnnotationLevel =
    | 'notice'
    | 'warning'
    | 'failure';
