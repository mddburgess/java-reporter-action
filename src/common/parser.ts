import * as core from "@actions/core";
import * as fs from "fs";
import { Saxophone } from "saxophone-ts";
import {
  CDATANode,
  TagCloseNode,
  TagOpenNode,
  TextNode,
} from "saxophone-ts/dist/types/src/static/nodes";

export default abstract class ReportParser<T> {
  private readonly context: string[] = ["root"];
  private readonly parser = new Saxophone()
    .on("tagOpen", (tag) => this.handleTagOpen(tag))
    .on("tagClose", (tag) => this.handleTagClose(tag))
    .on("text", (tag) => this.handleText(tag))
    .on("cdata", (tag) => this.handleText(tag));

  protected constructor(protected readonly report: T, private readonly reportPath: string) {}

  read() {
    try {
      const xml = fs.readFileSync(this.reportPath, { encoding: "utf-8" });
      this.parser.parse(xml);
      return this.report;
    } catch (error) {
      core.warning(`Failed to read report: ${this.reportPath}`);
      return undefined;
    }
  }

  protected getContext() {
    return this.context[0];
  }

  private handleTagOpen(tag: TagOpenNode) {
    this.onTagOpen(tag);
    if (tag.isSelfClosing) {
      this.onTagClose(tag);
    } else {
      this.context.unshift(tag.name);
    }
  }

  private handleTagClose(tag: TagCloseNode) {
    this.context.shift();
    this.onTagClose(tag);
  }

  private handleText(tag: TextNode | CDATANode) {
    this.onText(tag);
  }

  protected abstract onTagOpen(tag: TagOpenNode): void;
  protected abstract onTagClose(tag: TagCloseNode): void;
  protected abstract onText(tag: TextNode | CDATANode): void;
}
