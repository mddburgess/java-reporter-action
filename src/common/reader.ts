import * as core from '@actions/core';
import * as fs from 'fs';
import {Saxophone} from 'saxophone-ts';
import {CDATANode, TagCloseNode, TagOpenNode, TextNode} from 'saxophone-ts/dist/types/src/static/nodes';

abstract class ReportReader<T> {

    private readonly parser = new Saxophone()
        .on('tagOpen', tag => this.handleTagOpen(tag))
        .on('tagClose', tag => this.handleTagClose(tag))
        .on('text', tag => this.handleText(tag))
        .on('cdata', tag => this.handleText(tag));
    private readonly context: string[] = ['root'];
    protected report?: T;

    public readReport(reportPath: string): T | undefined {
        try {
            const xml = fs.readFileSync(reportPath, {encoding: 'utf-8'});
            this.parser.parse(xml);
            return this.report;
        } catch (error) {
            core.warning(`Failed to read report: ${reportPath}`);
            console.log(error);
            return undefined;
        }
    }

    protected getContext(): string {
        return this.context[0];
    }

    private handleTagOpen(tag: TagOpenNode): void {
        this.onTagOpen(tag);
        if (tag.isSelfClosing) {
            this.onTagClose(tag);
        } else {
            this.context.unshift(tag.name);
        }
    }

    private handleTagClose(tag: TagCloseNode): void {
        this.context.shift();
        this.onTagClose(tag);
    }

    private handleText(tag: TextNode | CDATANode): void {
        this.onText(tag);
    }

    protected abstract onTagOpen(tag: TagOpenNode): void;
    protected abstract onTagClose(tag: TagCloseNode): void;
    protected abstract onText(tag: TextNode | CDATANode): void;
}

export default ReportReader;
