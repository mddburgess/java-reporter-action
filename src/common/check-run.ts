import * as core from '@actions/core';
import * as github from '@actions/github';
import {Annotation} from './github';

class CheckRun {

    private readonly token;
    private readonly octokit;

    private readonly name;
    private checkRunId?: number;

    constructor(name: string) {
        this.token = core.getInput('github-token', {required: true});
        this.octokit = github.getOctokit(this.token);
        this.name = name;
    }

    public async conclude(request: CheckRunRequest) {
        if (request.output) {
            await this.createOrUpdateCheckRun({
                status: 'completed',
                conclusion: request.conclusion,
                output: {
                    title: request.output.title,
                    summary: request.output.summary
                }
            });
            await this.annotate(request.output);
        } else {
            await this.createOrUpdateCheckRun({
                status: 'completed',
                conclusion: request.conclusion
            });
        }
    }

    private async annotate(output: CheckOutput) {
        if (!output.annotations || output.annotations.length === 0) {
            return;
        }
        for (let i = 0; i < output.annotations.length; i += 50) {
            await this.updateCheckRun({
                output: {
                    title: output.title,
                    summary: output.summary,
                    annotations: output.annotations.slice(i, i + 50)
                }
            });
        }
    }

    private async createOrUpdateCheckRun(request: CheckRunRequest) {
        if (this.checkRunId === undefined) {
            await this.createCheckRun(request);
        } else {
            await this.updateCheckRun(request);
        }
    }

    private async createCheckRun(request: CheckRunRequest) {
        core.debug(`Creating ${this.name} check run...`);
        const response = await this.octokit.checks.create({
            ...github.context.repo,
            name: this.name,
            head_sha: this.resolveHeadSha(),
            ...request
        });
        this.checkRunId = response.data.id;
    }

    private async updateCheckRun(request: CheckRunRequest) {
        if (this.checkRunId === undefined) {
            throw Error('Cannot update a check run before creating it');
        }

        core.debug(`Updating ${this.name} check run...`);
        await this.octokit.checks.update({
            ...github.context.repo,
            check_run_id: this.checkRunId,
            ...request
        });
    }

    private resolveHeadSha() {
        return github.context.payload.pull_request
            ? github.context.payload.pull_request.head.sha
            : github.context.sha;
    }
}

interface CheckRunRequest {
    status?: CheckStatus,
    conclusion?: CheckConclusion,
    output?: CheckOutput
}

type CheckStatus =
    | 'queued'
    | 'in_progress'
    | 'completed';

type CheckConclusion =
    | 'success'
    | 'failure'
    | 'neutral'
    | 'cancelled'
    | 'skipped'
    | 'timed_out'
    | 'action_required';

interface CheckOutput {
    title: string,
    summary: string,
    text?: string,
    annotations?: Annotation[];
}

export default CheckRun;
