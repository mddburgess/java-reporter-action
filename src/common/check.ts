import * as core from '@actions/core';
import * as github from '@actions/github';
import * as glob from '@actions/glob';
import path from 'path';
import {getInput} from '@actions/core';

abstract class Check {

    private static readonly workspacePath = process.env.GITHUB_WORKSPACE || '';

    private readonly reportType: string;
    private readonly checkCondition: CheckCondition;
    private checkRunId?: number;

    protected constructor(reportType: string) {
        this.reportType = reportType;
        this.checkCondition = this.resolveCheckCondition();
    }

    private resolveCheckCondition() {
        const condition = getInput(this.reportType);
        switch (condition) {
            case 'required':
            case 'expected':
            case 'disabled':
                return CheckCondition[condition];
            case undefined:
                return CheckCondition.autodetect;
            default:
                core.warning(`Input '${this.reportType}' is invalid - must be one of ['required','expected','disabled']. Defaulting to 'expected'.`);
                return CheckCondition.expected;
        }
    }

    public async run() {
        const reportPaths = await this.findReports();

        const token = core.getInput('github-token', {required: true});
        const octokit = github.getOctokit(token);

        if (reportPaths.length > 0 || this.checkCondition >= CheckCondition.expected) {
            const response = await octokit.checks.create({
                ...github.context.repo,
                name: this.reportType,
                head_sha: github.context.payload.pull_request
                    ? github.context.payload.pull_request.head.sha
                    : github.context.sha,
                status: 'completed',
                conclusion: reportPaths.length ? 'success' :
                    this.checkCondition === CheckCondition.required ? 'failure' : 'skipped'
            });
            this.checkRunId = response.data.id;
        }

        core.info(`${this.reportType} check finished.`);
    }

    private async findReports() {
        core.startGroup(`Searching for ${this.reportType} reports...`);

        const searchPaths = this.reportSearchPaths();
        searchPaths.forEach(searchPath => core.info(searchPath));
        core.endGroup();

        const globber = await glob.create(searchPaths.join('\n'));
        const reportPaths = await globber.glob();

        core.startGroup(`Found ${reportPaths.length} ${this.reportType} reports.`);
        reportPaths.forEach(reportPath => core.info(path.relative(Check.workspacePath, reportPath)));
        core.endGroup();

        return reportPaths;
    }

    protected abstract reportSearchPaths(): string[];
}

enum CheckCondition {
    disabled,
    autodetect,
    expected,
    required
}

export default Check;
