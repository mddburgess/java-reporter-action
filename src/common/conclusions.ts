import { AnnotationSummary } from "../github/check-run";
import { CheckConclusion } from "../github/types";

const failureOnWarnings = (result: AnnotationSummary): CheckConclusion => {
  if (result.failures + result.warnings > 0) {
    return "failure";
  }
  return "success";
};

const neutralOnWarnings = (result: AnnotationSummary): CheckConclusion => {
  if (result.failures > 0) {
    return "failure";
  }
  if (result.warnings > 0) {
    return "neutral";
  }
  return "success";
}

export {
  failureOnWarnings,
  neutralOnWarnings
};
