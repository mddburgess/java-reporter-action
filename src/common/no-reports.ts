import CheckResult from "../check/result";
import { CheckAnnotation, CheckConclusion } from "../github/types";
import { RunCondition } from "../check";

export default class NoReportsResult extends CheckResult {
  constructor(
    private readonly friendlyName: string,
    private readonly runCondition: RunCondition,
    private readonly searchPaths: string[]
  ) {
    super();
  }

  shouldCompleteCheck(): boolean {
    return this.runCondition >= RunCondition.expected;
  }

  get conclusion(): CheckConclusion {
    return this.runCondition === RunCondition.required ? "failure" : "skipped";
  }

  get title(): string {
    return "No reports found";
  }

  get summary(): string {
    const runConditionName = RunCondition.required ? "required" : "expected";
    return `The ${this.friendlyName} check is ${runConditionName}, but no ${this.friendlyName} reports were found.`;
  }

  get text(): string | undefined {
    return ["### Search paths", "```sh", ...this.searchPaths, "```"].join("\n");
  }

  get annotations(): CheckAnnotation[] | undefined {
    return undefined;
  }
}
