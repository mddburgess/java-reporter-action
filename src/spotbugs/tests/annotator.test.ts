import {findFile} from '../../common/files';
import SpotbugsReportReader from '../reader';
import SpotbugsAnnotator from '../annotator';

describe('SpotBugs annotator', () => {

    jest.setTimeout(10000);

    it('can annotate a SpotBugs report', async () => {
        const reportPath = await findFile('**/spotbugsXml.xml');
        const report = new SpotbugsReportReader().readReport(reportPath) || fail();
        const annotations = await new SpotbugsAnnotator().annotate(report);
        expect(annotations).toMatchSnapshot();
    });
});
