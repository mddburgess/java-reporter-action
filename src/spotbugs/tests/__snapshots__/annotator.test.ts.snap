// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpotBugs annotator can annotate a SpotBugs report 1`] = `
Array [
  Object {
    "annotation_level": "failure",
    "end_line": 142,
    "message": "Impossible downcast of toArray() result to Integer[] in org.example.cpd.ErrorProne.classCastExceptionWithToArray()",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 142,
    "title": "Correctness: Impossible downcast of toArray() result",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 142,
    "message": "Dead store to array in org.example.cpd.ErrorProne.classCastExceptionWithToArray()",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 142,
    "title": "Dodgy code: Dead store to local variable",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 116,
    "message": "BigDecimal constructed from 1.123 in org.example.cpd.ErrorProne.avoidDecimalLiteralsInBigDecimalConstructor()",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 116,
    "title": "Correctness: BigDecimal constructed from double that isn't represented precisely",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 37,
    "message": "Primitive boxed just to call toString in org.example.cpd.ErrorProne.unnecessaryConversionTemporary(int)",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 37,
    "title": "Performance: Method allocates a boxed primitive just to call toString",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 37,
    "message": "org.example.cpd.ErrorProne.unnecessaryConversionTemporary(int) invokes inefficient new Integer(int) constructor; use Integer.valueOf(int) instead",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 37,
    "title": "Performance: Method invokes inefficient Number constructor; use static valueOf instead",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 89,
    "message": "Empty synchronized block in org.example.cpd.ErrorProne.emptySynchronizedBlock()",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 89,
    "title": "Multithreaded correctness: Empty synchronized block",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 100,
    "message": "There is an apparent infinite loop in org.example.cpd.ErrorProne.jumbledIncrementer()",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 100,
    "title": "Correctness: An apparent infinite loop",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 128,
    "message": "Null pointer dereference of string in org.example.cpd.ErrorProne.brokenNullCheck(String)",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 128,
    "title": "Correctness: Null pointer dereference",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 128,
    "message": "Load of known null value in org.example.cpd.ErrorProne.brokenNullCheck(String)",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 128,
    "title": "Dodgy code: Load of known null value",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 137,
    "message": "org.example.cpd.ErrorProne.checkSkipResult() may fail to clean up java.io.InputStream",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 137,
    "title": "Experimental: Method may fail to clean up stream or resource",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 135,
    "message": "org.example.cpd.ErrorProne.checkSkipResult() may fail to close stream",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 135,
    "title": "Bad practice: Method may fail to close stream",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 100,
    "message": "Complicated, subtle or wrong increment in for-loop org.example.cpd.ErrorProne.jumbledIncrementer()",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 100,
    "title": "Dodgy code: Complicated, subtle or wrong increment in for-loop",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 14,
    "message": "Nullcheck of string at line 14 of value previously dereferenced in org.example.cpd.ErrorProne.misplacedNullCheck(String)",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 14,
    "title": "Correctness: Nullcheck of value previously dereferenced",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 46,
    "message": "Return value of java.math.BigDecimal.add(BigDecimal) ignored in org.example.cpd.ErrorProne.uselessOperationOnImmutable()",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 46,
    "title": "Correctness: Method ignores return value",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 136,
    "message": "org.example.cpd.ErrorProne.checkSkipResult() ignores result of java.io.FileInputStream.skip(long)",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 136,
    "title": "Bad practice: Method ignores results of InputStream.skip()",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 140,
    "message": "Useless object stored in variable collection of method org.example.cpd.ErrorProne.classCastExceptionWithToArray()",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 140,
    "title": "Dodgy code: Useless object created",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 113,
    "message": "Method org.example.cpd.ErrorProne.avoidBranchingStatementAsLastInLoop() seems to be useless",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 113,
    "title": "Dodgy code: Useless non-empty void method",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 143,
    "message": "Method org.example.cpd.ErrorProne.classCastExceptionWithToArray() seems to be useless",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 143,
    "title": "Dodgy code: Useless non-empty void method",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 53,
    "message": "Method org.example.cpd.ErrorProne.dontUseFloatTypeForLoopIndices() seems to be useless",
    "path": "example/src/main/java/org/example/cpd/ErrorProne.java",
    "start_line": 53,
    "title": "Dodgy code: Useless non-empty void method",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 16,
    "message": "Null pointer dereference of conn in org.example.pmd.BestPractices.checkResultSet()",
    "path": "example/src/main/java/org/example/pmd/BestPractices.java",
    "start_line": 16,
    "title": "Correctness: Null pointer dereference",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 19,
    "message": "org.example.pmd.BestPractices.checkResultSet() may fail to clean up java.sql.ResultSet",
    "path": "example/src/main/java/org/example/pmd/BestPractices.java",
    "start_line": 19,
    "title": "Experimental: Method may fail to clean up stream or resource",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 19,
    "message": "org.example.pmd.BestPractices.checkResultSet() may fail to clean up java.sql.Statement",
    "path": "example/src/main/java/org/example/pmd/BestPractices.java",
    "start_line": 19,
    "title": "Experimental: Method may fail to clean up stream or resource",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 16,
    "message": "org.example.pmd.BestPractices.checkResultSet() may fail to close Statement",
    "path": "example/src/main/java/org/example/pmd/BestPractices.java",
    "start_line": 16,
    "title": "Bad practice: Method may fail to close database resource",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 11,
    "message": "Unread public/protected field: org.example.pmd.BestPractices.ipAddress",
    "path": "example/src/main/java/org/example/pmd/BestPractices.java",
    "start_line": 11,
    "title": "Dodgy code: Unread public/protected field",
  },
  Object {
    "annotation_level": "warning",
    "end_line": NaN,
    "message": "Unused field: org.example.pmd.BestPractices.unusedPrivateField",
    "path": "example/src/main/java/org/example/pmd/BestPractices.java",
    "start_line": NaN,
    "title": "Performance: Unused field",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 41,
    "message": "Dead store to instance in org.example.pmd.CodeStyle.uselessQualifiedThis()",
    "path": "example/src/main/java/org/example/pmd/CodeStyle.java",
    "start_line": 41,
    "title": "Dodgy code: Dead store to local variable",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 8,
    "message": "Dead store to x in org.example.pmd.Design.collapsibleIfStatements(boolean, boolean)",
    "path": "example/src/main/java/org/example/pmd/Design.java",
    "start_line": 8,
    "title": "Dodgy code: Dead store to local variable",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 51,
    "message": "Impossible downcast of toArray() result to Integer[] in org.example.pmd.ErrorProne.classCastExceptionWithToArray()",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 51,
    "title": "Correctness: Impossible downcast of toArray() result",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 51,
    "message": "Dead store to array in org.example.pmd.ErrorProne.classCastExceptionWithToArray()",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 51,
    "title": "Dodgy code: Dead store to local variable",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 25,
    "message": "BigDecimal constructed from 1.123 in org.example.pmd.ErrorProne.avoidDecimalLiteralsInBigDecimalConstructor()",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 25,
    "title": "Correctness: BigDecimal constructed from double that isn't represented precisely",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 135,
    "message": "Primitive boxed just to call toString in org.example.pmd.ErrorProne.unnecessaryConversionTemporary(int)",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 135,
    "title": "Performance: Method allocates a boxed primitive just to call toString",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 135,
    "message": "org.example.pmd.ErrorProne.unnecessaryConversionTemporary(int) invokes inefficient new Integer(int) constructor; use Integer.valueOf(int) instead",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 135,
    "title": "Performance: Method invokes inefficient Number constructor; use static valueOf instead",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 94,
    "message": "Empty synchronized block in org.example.pmd.ErrorProne.emptySynchronizedBlock()",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 94,
    "title": "Multithreaded correctness: Empty synchronized block",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 105,
    "message": "There is an apparent infinite loop in org.example.pmd.ErrorProne.jumbledIncrementer()",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 105,
    "title": "Correctness: An apparent infinite loop",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 37,
    "message": "Null pointer dereference of string in org.example.pmd.ErrorProne.brokenNullCheck(String)",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 37,
    "title": "Correctness: Null pointer dereference",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 37,
    "message": "Load of known null value in org.example.pmd.ErrorProne.brokenNullCheck(String)",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 37,
    "title": "Dodgy code: Load of known null value",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 46,
    "message": "org.example.pmd.ErrorProne.checkSkipResult() may fail to clean up java.io.InputStream",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 46,
    "title": "Experimental: Method may fail to clean up stream or resource",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 44,
    "message": "org.example.pmd.ErrorProne.checkSkipResult() may fail to close stream",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 44,
    "title": "Bad practice: Method may fail to close stream",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 105,
    "message": "Complicated, subtle or wrong increment in for-loop org.example.pmd.ErrorProne.jumbledIncrementer()",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 105,
    "title": "Dodgy code: Complicated, subtle or wrong increment in for-loop",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 112,
    "message": "Nullcheck of string at line 112 of value previously dereferenced in org.example.pmd.ErrorProne.misplacedNullCheck(String)",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 112,
    "title": "Correctness: Nullcheck of value previously dereferenced",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 144,
    "message": "Return value of java.math.BigDecimal.add(BigDecimal) ignored in org.example.pmd.ErrorProne.uselessOperationOnImmutable()",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 144,
    "title": "Correctness: Method ignores return value",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 45,
    "message": "org.example.pmd.ErrorProne.checkSkipResult() ignores result of java.io.FileInputStream.skip(long)",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 45,
    "title": "Bad practice: Method ignores results of InputStream.skip()",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 49,
    "message": "Useless object stored in variable collection of method org.example.pmd.ErrorProne.classCastExceptionWithToArray()",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 49,
    "title": "Dodgy code: Useless object created",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 22,
    "message": "Method org.example.pmd.ErrorProne.avoidBranchingStatementAsLastInLoop() seems to be useless",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 22,
    "title": "Dodgy code: Useless non-empty void method",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 52,
    "message": "Method org.example.pmd.ErrorProne.classCastExceptionWithToArray() seems to be useless",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 52,
    "title": "Dodgy code: Useless non-empty void method",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 58,
    "message": "Method org.example.pmd.ErrorProne.dontUseFloatTypeForLoopIndices() seems to be useless",
    "path": "example/src/main/java/org/example/pmd/ErrorProne.java",
    "start_line": 58,
    "title": "Dodgy code: Useless non-empty void method",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 18,
    "message": "Possible doublecheck on org.example.pmd.Multithreading.object in org.example.pmd.Multithreading.doubleCheckedLocking()",
    "path": "example/src/main/java/org/example/pmd/Multithreading.java",
    "start_line": 16,
    "title": "Multithreaded correctness: Possible double check of field",
  },
  Object {
    "annotation_level": "failure",
    "end_line": 8,
    "message": "Dead store to tg in org.example.pmd.Multithreading.avoidThreadGroup()",
    "path": "example/src/main/java/org/example/pmd/Multithreading.java",
    "start_line": 8,
    "title": "Dodgy code: Dead store to local variable",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 19,
    "message": "Inconsistent synchronization of org.example.pmd.Multithreading.object; locked 40% of time",
    "path": "example/src/main/java/org/example/pmd/Multithreading.java",
    "start_line": 19,
    "title": "Multithreaded correctness: Inconsistent synchronization",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 12,
    "message": "org.example.pmd.Multithreading.dontCallThreadRun() explicitly invokes run on a thread (did you mean to start it instead?)",
    "path": "example/src/main/java/org/example/pmd/Multithreading.java",
    "start_line": 12,
    "title": "Multithreaded correctness: Invokes run on a thread (did you mean to start it instead?)",
  },
  Object {
    "annotation_level": "warning",
    "end_line": 12,
    "message": "org.example.pmd.Performance.booleanInstantiation() invokes inefficient Boolean constructor; use Boolean.valueOf(...) instead",
    "path": "example/src/main/java/org/example/pmd/Performance.java",
    "start_line": 12,
    "title": "Performance: Method invokes inefficient Boolean constructor; use Boolean.valueOf(...) instead",
  },
]
`;
